# In this notebbok we tried to make a Colaborative Filtering Engine by nearest neightbors fitting but we decide to take other way because of the data sparsity

In [1]:
%matplotlib inline
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd

In [2]:
df = pd.read_csv('df_to_recommender.csv')

In [3]:
df=df[['user_id', 'Id','overall']].copy()

In [4]:
df.shape

(7771, 3)

### We start trying with rating vaues over/equal 4.0 it is what dataset has

In [5]:
def pivotarray(df):
    users,i= np.unique(df['user_id'],return_inverse=True)
    item,j= np.unique(df['Id'],return_inverse=True)
    a=np.zeros((len(users),len(item)))
    a[i,j]=df['overall']
    return a

In [6]:
ratings1 = pivotarray (df)
ratings1

array([[ 0.,  0.,  0., ...,  0.,  0.,  0.],
       [ 0.,  0.,  0., ...,  0.,  0.,  0.],
       [ 0.,  0.,  0., ...,  0.,  0.,  0.],
       ..., 
       [ 0.,  0.,  0., ...,  0.,  0.,  0.],
       [ 0.,  0.,  0., ...,  0.,  0.,  0.],
       [ 0.,  0.,  0., ...,  0.,  0.,  0.]])

In [7]:
sparsity = float(len(ratings1.nonzero()[0])) 
sparsity /= (ratings1.shape[0] * ratings1.shape[1]) 
sparsity *= 100 
print('Sparsity: {:4.2f}%'.format(sparsity)) 
 

Sparsity: 0.47%


In [8]:
df2 = df[df.overall>4.0]

In [9]:
df2.shape

(2542, 3)

In [10]:
ratings2 = pivotarray (df2)

In [11]:
sparsity = float(len(ratings2.nonzero()[0])) 
sparsity /= (ratings2.shape[0] * ratings2.shape[1]) 
sparsity *= 100 
print('Sparsity: {:4.2f}%'.format(sparsity)) 

Sparsity: 0.49%


In [12]:
df3 = df[df.overall>=5.0]

In [13]:
df3.shape

(379, 3)

In [14]:
ratings3 = pivotarray (df3)

### As it can be seen, taking only 379 rows we get 0.89% sparsity. It means hat we only have rating information for 0.89% the data and for the others it is just zeros

In [15]:
sparsity = float(len(ratings3.nonzero()[0])) 
sparsity /= (ratings3.shape[0] * ratings3.shape[1]) 
sparsity *= 100 
print('Sparsity: {:4.2f}%'.format(sparsity)) 

Sparsity: 0.89%
